src/main/java/com/redhat/pantheon/model/api/SlingResourceAdapterFactory.java

```diff
@@ -0,0 +1,65 @@
+package com.redhat.pantheon.model.api;
+
+import org.apache.sling.api.adapter.AdapterFactory;
+import org.apache.sling.api.resource.Resource;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.osgi.service.component.annotations.Component;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.reflect.InvocationTargetException;
+
+import static org.apache.sling.api.adapter.AdapterFactory.ADAPTABLE_CLASSES;
+import static org.apache.sling.api.adapter.AdapterFactory.ADAPTER_CLASSES;
+
+/**
+ * An adapter factory for model objects of type SlingResource. This component makes sure that
+ * calls to {@link Resource#adaptTo(Class)} with a subtype of {@link SlingResource} as a parameter
+ * will return an object.
+ *
+ * Types need to registered as adapter classes by simply adding a property in the @{@link Component}
+ * definition at the top of this class.
+ */
+@Component(
+        name = "Pantheon - JCRModel Adapter Factory",
+        service = AdapterFactory.class,
+        property = {
+                ADAPTABLE_CLASSES + "=org.apache.sling.api.resource.Resource",
+                ADAPTER_CLASSES + "=com.redhat.pantheon.model.Module",
+        }
+)
+public class SlingResourceAdapterFactory implements AdapterFactory {
+
+    private static final Logger log = LoggerFactory.getLogger(SlingResourceAdapterFactory.class);
+
+    @Override
+    public <AdapterType> @Nullable AdapterType getAdapter(@NotNull Object adaptable, @NotNull Class<AdapterType> type) {
+
+        // ensure the adapter type is a subclass of SlingResource
+        if(!SlingResource.class.isAssignableFrom(type)) {
+            log.error("Error in " + this.getClass().getSimpleName() + ": adapter type not a subclass of SlingResource ("
+                    + type.getName() + ")");
+            return null;
+        }
+
+        // Ensure the adaptable object is of an appropriate type
+        if (!(adaptable instanceof Resource) || (adaptable == null)) {
+            log.error("Trying to adapt object {" + adaptable + "} to SlingResource, but object is null or not of type Resource");
+            return null;
+        }
+
+        Resource resource = ((Resource) adaptable);
+
+        // the adapter type (sublass of SlingResource) should have a one arg constructor which takes a resource
+        AdapterType adapter = null;
+        try {
+            adapter = type.getConstructor(Resource.class)
+                    .newInstance(resource);
```

Comment:
Yes, I'm not entirely happy with this approach as it requires a one-arg constructor. With some use of reflection we can avoid this, but I'm happy to leave that improvement for another time.

Reply:
What are you unhappy with? I guess now that I'm taking another look at this, I'm confused by the syntax. Why have this:
`adapter = type.getConstructor(Resource.class).newInstance(resource);`
...when you can have this...
`adapter = new Module(resource);`
...?

I guess the advantage of the first approach would be that it continues to cover all cases when we add more classes than just Module, but the point I'm trying to make is that `adapter` appears to just be the exact instance of the thing you're adapting TO, so you have some flexibility in how you create that thing. I don't see a one-arg constructor as a particular requirement here.

Reply:
You are correct: the reflection code is used so that it applies to any subclass of `SlingResource` which needs to have the one-arg constructor (being called there). My point is that since we are doing reflection, we could just as well look for the `resource` property inside the `SlingResource` object (which we know will be there) and "inject" it. There is in fact no need for the one-arg constructor, but I would rather push that out as an improvement for the next revision.